File: kt/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/IfStatementSyntax.kt
package dev.phytolizer.minsk.analysis.syntax

class IfStatementSyntax(
    val ifKeyword: SyntaxToken,
    val condition: ExpressionSyntax,
    val thenStatement: StatementSyntax,
    val elseClause: ElseClauseSyntax?,
) : StatementSyntax() {
    override val kind: SyntaxKind
        get() = SyntaxKind.IfStatement
    override val children: List<SyntaxNode>
        get() = listOfNotNull(ifKeyword, condition, thenStatement, elseClause)
}

File: cs/Minsk/CodeAnalysis/Syntax/IfStatementSyntax.cs
namespace Minsk.CodeAnalysis.Syntax;

public sealed class IfStatementSyntax : StatementSyntax
{
    public IfStatementSyntax(SyntaxToken ifKeyword, ExpressionSyntax condition, StatementSyntax thenStatement, ElseClauseSyntax? elseClause)
    {
        IfKeyword = ifKeyword;
        Condition = condition;
        ThenStatement = thenStatement;
        ElseClause = elseClause;
    }

    public override SyntaxKind Kind => SyntaxKind.IfStatement;

    public override IEnumerable<SyntaxNode> Children
    {
        get
        {
            yield return IfKeyword;
            yield return Condition;
            yield return ThenStatement;
            if (ElseClause != null)
            {
                yield return ElseClause;
            }
        }
    }

    public SyntaxToken IfKeyword { get; }
    public ExpressionSyntax Condition { get; }
    public StatementSyntax ThenStatement { get; }
    public ElseClauseSyntax? ElseClause { get; }
}
